R/archive.R

```diff
@@ -377,6 +354,63 @@ new_epi_archive <- function(
   )
 }
 
+#' given a tibble as would be found in an epi_archive, remove duplicate entries.
+#' @description
+#' works by shifting all rows except the version, then comparing values to see
+#'   if they've changed. We need to arrange in descending order, but note that
+#'   we don't need to group, since at least one column other than version has
+#'   changed, and so is kept.
+#' @keywords internal
+#' @importFrom dplyr filter
+compactify_tibble <- function(df, keys) {
+  df %>%
+    arrange(!!!keys) %>%
+    filter(if_any(
+      c(everything(), -version), # all non-version columns
+      ~ !is_locf(.)
+    ))
+}
+
+#' get the entries that `compactify_tibble` would remove
+#' @keywords internal
+#' @importFrom dplyr filter if_all everything
+removed_by_compactify <- function(df, keys) {
+  df %>%
+    arrange(!!!keys) %>%
+    filter(if_all(
+      c(everything(), -version),
+      ~ is_locf(.)
+    )) # nolint: object_usage_linter
+}
+
+#' Checks to see if a value in a vector is LOCF
+#' @description
+#' LOCF meaning last observation carried forward. lags the vector by 1, then
+#'   compares with itself. For floats it uses float comparison, otherwise it
+#'   uses equality.
+#' `NA`'s are considered equal to `NA`'s, while nan's are not.
+#' @importFrom dplyr lag if_else near
+#' @keywords internal
+is_locf <- function(vec) { # nolint: object_usage_linter
+  lag_vec <- lag(vec)
+  if (typeof(vec) == "double") {
+    if_else(
+      !is.na(vec) & !is.na(lag_vec),
+      near(vec, lag_vec),
+      is.na(vec) & is.na(lag_vec)
+    ) %>%
+      return()
```

Comment:
issue: `%>% return()` doesn't actually return!  It appears to be effectively `%>% identity()`.  This may mislead us later.

```r
library(magrittr)
f <- function(x) {x %>% return(); 42}
f(1)
#> [1] 42
```

```suggestion
    )
```
or do `result <- ......` + `return(result)` so we can keep the convenience of Ctrl-F return.

Reply:
good catch, weird that things were working anyways.